### PR TITLE
Added '-a' (--append) arg to append a line of text to the notes file.

### PR DESCRIPTION
**Pull request by dvklopfenstein**
Please consider this pull request, which augments the 'pubs note' command so text can be appended to the note file without needed to open the file in an editor, write it, then close it.

### CODE REVIEW

File: pubs/commands/note_cmd.py [hunk 1]
```diff
@@ -18,7 +19,12 @@ def command(conf, args):
     ui = get_ui()
     rp = repo.Repository(conf)
     citekey = resolve_citekey(rp, args.citekey, ui=ui, exit_on_fail=True)
+    append = args.append
```
Comment (benureau) [inline pubs/commands/note_cmd.py]: With append being a common method for lists, I guess we could find a better name for this variable.

File: pubs/commands/note_cmd.py [hunk 2]
```diff
@@ -18,7 +19,12 @@ def command(conf, args):
...
     notepath = rp.databroker.real_notepath(citekey, rp.conf['main']['note_extension'])
-    ui.edit_file(notepath, temporary=False)
+    if append is None:
+        ui.edit_file(notepath, temporary=False)
+    else:
+        with open(notepath, 'a') as fd:
```
Comment (benureau) [inline pubs/commands/note_cmd.py]: With Py27 support still ongoing, and encoding problems galore,  here you should use this function instead:
https://github.com/pubs/pubs/blob/5551e8c04b21fd1ba5a45a204f25d19ab7727462/pubs/content.py#L98

File: tests/test_note_append.py [hunk 3]
```diff
@@ -34,11 +34,11 @@ def test_note_edit(self):
         self.execute_cmds(cmds)
         note_lines.append('bbb')
         self.assertFileContentEqual(fin_notes, self._get_note_content(note_lines))
-        # Test adding Japanese character
-        cmds = [('pubs note Page99 -a ソ')]
-        self.execute_cmds(cmds)
-        note_lines.append('ソ')
-        self.assertFileContentEqual(fin_notes, self._get_note_content(note_lines))
+        # # Test adding Japanese character
```
[the remark this answers is not in the document]
Comment (dvklopfenstein) [reply tests/test_note_append.py]: I would very much appreciate you taking a crack at this after the merge. In my work, I rarely work with anything but the basic 256 ASCII set.

File: tests/test_note_append.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    def test_note_append(self):
+        """Test appending the note file using the command-line argument, -a"""
+        fin_notes = os.path.join(self.note_dir, 'Page99.txt')
+        # Test adding first line
+        cmds = [('pubs note Page99 -a aaa')]
+        self.execute_cmds(cmds)
+        note_lines = ['aaa']
+        self.assertFileContentEqual(fin_notes, self._get_note_content(note_lines))
+        # Test adding additional line
+        cmds = [('pubs note Page99 -a bbb')]
```
[the remark this answers is not in the document]
Comment (dvklopfenstein) [reply tests/test_note_append.py]: Appending a multi-word note line worked great on the command-line:
```
$ pubs note Subramanian_2005 -a 'aaa bbb'

$ tail $(find pubs/notes/ -name Subramanian_2005.\*)
proj_dis_clu/doc/papers/diffexpress/y2005_17448_GSEA_Subramanian_Mesirov.pdf
aaa bbb
```

But when I tried to add appending 'aaa bbb' to the list of tests, I see this error:
```
======================================================================
ERROR: test_note_append (__main__.TestNoteAppend)
Test appending the note file using the command-line argument, -a
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/cygdrive/c/Users/note2/Data/git/pubs/pubs/pubs_cmd.py", line 104, in execute
    args.func(conf, args)
  File "/cygdrive/c/Users/note2/Data/git/pubs/pubs/commands/init_cmd.py", line 48, in command
    config.save_conf(conf)
  File "/cygdrive/c/Users/note2/Data/git/pubs/pubs/config/conf.py", line 78, in save_conf
    with open(conf.filename, 'wb') as f:
  File "/usr/lib/python3.6/site-packages/pyfakefs/fake_filesystem.py", line 4640, in open
    newline, closefd, opener)
  File "/usr/lib/python3.6/site-packages/pyfakefs/fake_filesystem.py", line 5168, in __call__
    return self.call(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/pyfakefs/fake_filesystem.py", line 5252, in call
    apply_umask=True, raw_io=self.raw_io)
  File "/usr/lib/python3.6/site-packages/pyfakefs/fake_filesystem.py", line 2602, in create_file_internally
    error_fct(errno.ENOENT, parent_directory)
  File "/usr/lib/python3.6/site-packages/pyfakefs/fake_filesystem.py", line 995, in raise_io_error
    raise IOError(errno, self._error_message(errno), filename)
FileNotFoundError: [Errno 2] No such file or directory in the fake filesystem: '/cygdrive/c/Users/note2/Data/git/bibliography/pubs'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/cygdrive/c/Users/note2/Data/git/pubs/tests/test_usecase.py", line 131, in execute_cmds
    _, stdout, stderr = capture_wrap(actual_cmd.split())
  File "/cygdrive/c/Users/note2/Data/git/pubs/tests/fake_env.py", line 43, in newf
    return f(*args, **kwargs), _get_fake_stdio_ucontent(sys.stdout), _get_fake_stdio_ucontent(sys.stderr)
  File "/cygdrive/c/Users/note2/Data/git/pubs/pubs/pubs_cmd.py", line 107, in execute
    if not uis.get_ui().handle_exception(e):
  File "/cygdrive/c/Users/note2/Data/git/pubs/tests/fake_env.py", line 88, in handler
    original_exception_handler(ui, exc)
  File "/cygdrive/c/Users/note2/Data/git/pubs/pubs/uis.py", line 105, in handle_exception
    self.exit()
  File "/cygdrive/c/Users/note2/Data/git/pubs/pubs/uis.py", line 94, in exit
    sys.exit(error_code)
SystemExit: 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tests/test_note_append.py", line 21, in setUp
    self.execute_cmds(init)
  File "/cygdrive/c/Users/note2/Data/git/pubs/tests/test_usecase.py", line 151, in execute_cmds
    raise FakeSystemExit(*exc.args).with_traceback(tb)
  File "/cygdrive/c/Users/note2/Data/git/pubs/tests/test_usecase.py", line 131, in execute_cmds
    _, stdout, stderr = capture_wrap(actual_cmd.split())
  File "/cygdrive/c/Users/note2/Data/git/pubs/tests/fake_env.py", line 43, in newf
    return f(*args, **kwargs), _get_fake_stdio_ucontent(sys.stdout), _get_fake_stdio_ucontent(sys.stderr)
  File "/cygdrive/c/Users/note2/Data/git/pubs/pubs/pubs_cmd.py", line 107, in execute
    if not uis.get_ui().handle_exception(e):
  File "/cygdrive/c/Users/note2/Data/git/pubs/tests/fake_env.py", line 88, in handler
    original_exception_handler(ui, exc)
  File "/cygdrive/c/Users/note2/Data/git/pubs/pubs/uis.py", line 105, in handle_exception
    self.exit()
  File "/cygdrive/c/Users/note2/Data/git/pubs/pubs/uis.py", line 94, in exit
    sys.exit(error_code)
tests.test_usecase.FakeSystemExit: Exited with code: 1.

----------------------------------------------------------------------
Ran 1 test in 0.138s

FAILED (errors=1)
```

Upon investigation, it looks like this line in tests/test_usecase.py may be splitting up the words in a multiword note:
```
class CommandTestCase(fake_env.TestFakeFs):
    """Abstract TestCase intializing the fake filesystem."""

...
                        capture_wrap = fake_env.capture(pubs_cmd.execute,
                                                        verbose=PRINT_OUTPUT)
                        _, stdout, stderr = capture_wrap(actual_cmd.split())
...
```

I fiddled with the code  in tests/test_usecase.py, but was ended up spending too much time to add support of multiple lines in the *actual_cmd* variable. Would it be possible for someone more familiar with the *pubs* test code to add the ability to handle arguments having multiple words?

File: pubs/commands/note_cmd.py [hunk 5]
```diff
@@ -19,6 +22,10 @@ def command(conf, args):
     rp = repo.Repository(conf)
     citekey = resolve_citekey(rp, args.citekey, ui=ui, exit_on_fail=True)
     notepath = rp.databroker.real_notepath(citekey, rp.conf['main']['note_extension'])
-    ui.edit_file(notepath, temporary=False)
+    if args.append is None:
+        ui.edit_file(notepath, temporary=False)
+    else:
+        latestnote = '{TXT}\n'.format(TXT=args.append)
```
[the remark this answers is not in the document]
Comment (dvklopfenstein) [reply pubs/commands/note_cmd.py]: Updated for pep. Fyi, *pylint* did not catch this.

File: tests/test_note_append.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        # # Test adding Japanese character
+        # cmds = [('pubs note Page99 -a ソ')]
+        # self.execute_cmds(cmds)
+        # note_lines.append('ソ')
+        # self.assertFileContentEqual(fin_notes, self._get_note_content(note_lines))
+
+    @staticmethod
+    def _get_note_content(note_lines):
+        """Given a list of note lines, return full note file content"""
+        return '{LINES}\n'.format(LINES='\n'.join(note_lines))
```
[the remark this answers is not in the document]
Comment (dvklopfenstein) [reply tests/test_note_append.py]: Also updated...  Thank you for your comments.

**Comment by dvklopfenstein** [comment]
I was able to add the ***append a multi-word line to a bib note file*** by modifying the *test_usage* code.

The modifications are refactoring only. No real code has been changed.

With this change, a single pubs command is executed in its own small function, *execute_cmdsplit*. With this change, the command can be split in the test in a custom way instead of using always str.split().

File: tests/test_note_append.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dvklopfenstein) [reply tests/test_note_append.py]: NOTE: This test now works with minor test code refactoring 

Test appending a string with spaces:
```
pubs not Page99 -a "xxx yyy "
```

Refactoring is in this commit:
https://github.com/pubs/pubs/pull/201/commits/b75d77bf2499df607fb0f5fb8b04eb2718fd2165

**Comment by dvklopfenstein** [comment]
Appending a line of text to a bib note from the command line is functioning well in in my daily work.
 
Please let me know if there is anything else I should do. Thank you for this great software.

**Comment by benureau** [comment]
Thank you for this. Indeed, the multi-word did not work because how the tests were setup, I failed to see that; thanks for improving the code there also. I am merging this, and then I'll push a commit to fix the Python 2 situation. Thanks for improving pubs!

**Comment by benureau** [comment]
@dvklopfenstein: All right, the fix is in with #203, a bit more involved than I thought. I also added you to the contributors in the readme and updated the changelog.

**Pull request merged**